test/ConfigSpec.hs

```diff
@@ -172,6 +175,60 @@ spec = around_ (inTempDirectoryNamed "foo") $ do
       Right (_, c) <- readPackageConfig "package.yaml"
       packageExtraSourceFiles c `shouldBe` ["CHANGES.markdown", "README.markdown"]
 
+    it "accepts file globs in extra-source-files" $ do
+      writeFile "package.yaml" [i|
+        extra-source-files:
+          - res/*
+        |]
+      let files = map ("res" </>) ["foo.bar", "hello", "world"]
+      mapM_ touch files
+      Right (_, c) <- readPackageConfig "package.yaml"
+      packageExtraSourceFiles c `shouldBe` files
+
+    it "disallows duplicates in extra-source-files in presence of globs" $ do
+      writeFile "package.yaml" [i|
+        extra-source-files:
+          - res/*
+          - res/hello
+        |]
+      touch "res/hello"
+      Right (_, c) <- readPackageConfig "package.yaml"
+      packageExtraSourceFiles c `shouldBe` ["res/hello"]
+
+    it "expands directory globs" $ do
+      writeFile "package.yaml" [i|
+        extra-source-files:
+          - res/*/*
+        |]
+      touch "res/foo/hello.foo"
+      touch "res/bar/hello.bar"
+      Right (_, c) <- readPackageConfig "package.yaml"
+      packageExtraSourceFiles c `shouldBe`
+        ["res/bar/hello.bar", "res/foo/hello.foo"]
+
+    it "expands globs relative to the package.yaml file" $ do
+      writeFile "package.yaml" [i|
+        extra-source-files:
+          - res/*/*
+        |]
+      touch "res/foo/hello.foo"
+      touch "res/bar/hello.bar"
+      setCurrentDirectory "res/foo"
+      Right (_, c) <- readPackageConfig "../../package.yaml"
+      packageExtraSourceFiles c `shouldBe`
+        ["res/bar/hello.bar", "res/foo/hello.foo"]
+
+    it "warns if an extra-source-file doesn't match anything" $ do
+      writeFile "package.yaml" [i|
+        extra-source-files:
+          - missing.foo
+          - res/*
+        |]
+      fmap fst <$> readPackageConfig "package.yaml" `shouldReturn` Right [
+            "Specified extra-source-file \"missing.foo\" does not exist, leaving as-is"
+          , "Specified extra-source-file \"res/*\" does not exist, leaving as-is"
+          ]
```

Comment:
Do the non-matching patterns end up in the cabal file? What about dismissing them?